apps/elixir_ls_debugger/lib/debugger/server.ex

```diff
@@ -1297,4 +1280,63 @@ defmodule ElixirLS.Debugger.Server do
     server = Process.info(self())[:registered_name] || self()
     {__MODULE__, reason, [server]}
   end
+
+  defp update_threads(state = %__MODULE__{}) do
+    pids = :erlang.processes()
+    {state, thread_ids, new_ids} = ensure_thread_ids(state, pids)
+
+    for thread_id <- new_ids do
+      Output.send_event("thread", %{
+        "reason" => "started",
+        "threadId" => thread_id
+      })
+    end
+
+    exited_pids = Map.keys(state.threads_inverse) -- pids
+
+    state =
+      Enum.reduce(exited_pids, state, fn pid, state ->
+        handle_process_exit(state, pid)
+      end)
+
+    {schedule_update_threads(state), thread_ids}
+  end
+
+  defp handle_process_exit(state = %__MODULE__{}, pid) when is_pid(pid) do
+    {thread_id, threads_inverse} = state.threads_inverse |> Map.pop(pid)
+    paused_processes = remove_paused_process(state, pid)
+
+    state = %__MODULE__{
+      state
+      | threads: state.threads |> Map.delete(thread_id),
+        paused_processes: paused_processes,
+        threads_inverse: threads_inverse
+    }
+
+    if thread_id do
+      Output.send_event("thread", %{
+        "reason" => "exited",
+        "threadId" => thread_id
+      })
+    end
+
+    state
+  end
+
+  defp process_name(process_info) do
+    registered_name = Keyword.get(process_info, :registered_name)
+
+    if registered_name do
+      inspect(registered_name)
+    else
+      {mod, func, arity} = Keyword.fetch!(process_info, :initial_call)
+      "#{inspect(mod)}.#{to_string(func)}/#{arity}"
+    end
+  end
+
+  defp schedule_update_threads(state = %__MODULE__{update_threads_ref: old_ref}) do
+    if old_ref, do: Process.cancel_timer(old_ref, info: false)
+    ref = Process.send_after(self(), :update_threads, 3000)
```

Comment:
Is a thread id is just a mapping of some number to a process?
If so, why don't we just monitor the processes and clean up after one of them dies?

Reply:
> Is a thread id is just a mapping of some number to a process?

Yes, DAP requires that is an unique int. I was thinking if we can drop it altogether and just dump pid internal data representation. It's 28 bit uint IIRC but that won't be sufficient make it work in distributed scenario. Node name atom is part of pid.

> If so, why don't we just monitor the processes and clean up after one of them dies?

We monitor a process once it hits a breakpoint. I'm not convinced it is a good idea to monitor every process in the (possibly distributed) system. Besides that I don't know if it's possible to auto monitor process on start. You can auto break but that's not we are after.

Reply:
> I'm not convinced it is a good idea to monitor every process in the (possibly distributed) system. 

Can you elaborate on this? Monitoring processes is core to the BEAM and works very well over distribution. Discord, for example monitors millions of processes over dozens of systems with no problems. It's also entirely possible that I don't understand what's going on here.

> Besides that I don't know if it's possible to auto monitor process on start. 

In practice, it doesn't matter, once you have a pid (even a dead one) and you monitor it, you'll get the DOWN message immediately.

Reply:
> Can you elaborate on this?

I don't know the limits here. I was thinking that when processes die at rate higher than we can handle `:DOWN` it will DoS the server process

 > In practice, it doesn't matter, once you have a pid (even a dead one) and you monitor it, you'll get the DOWN message immediately.
 
 I know about dead pid behavior. How can I get the pid in the first place? I don't know of any way besides calling `:erlang.processes/1` or `:erlang:registered/1`. Is there a way to get notified whenever a process is spawned?

Reply:
I guess I really don't understand what's going on here at all. That said, I wouldn't be too concerned about excessive monitoring, as this is a language server and will likely not have more than a couple thousand processes at one time.

Is there somewhere I can read about how the debugger works?

Reply:
> Is there somewhere I can read about how the debugger works?

There's not much out there. The [:int](https://www.erlang.org/doc/man/int.html) module provides most of the functionality. Not everything is documented and some internal OTP stuff is used here. Fortunately it's been stable since Jake Becker did the original implementation. Besides that there's [DAP specification](https://microsoft.github.io/debug-adapter-protocol/)

The high level overview of debugger is:
1. Start DAP server and intercept IO (protocol is similar to LSP)
2. Receive config from the client
3. Call a few mix tasks (loadconfig, app.start)
4. Optionally require .exs files (e.g. tests)
5. Interpret modules with `:int.ni` and set breakpoints
6. Notify client and wait for start command
7. Call the mix task from config (e.g. mix test)
8. Sleep infintely (`Mix.Task.run` is non blocking and does not send events)
9. When a breakpoint is hit a callback is invoked which sends cast message to the server process
10. Server monitors debugged process and sends `stopped` event to the debugger client
11. When a process is stepped through the client requests stacktrace, variables (ass seen by the emulator so they don't match elixir ones) etc.